src/System.Private.CoreLib/shared/System/Threading/SpinWait.cs

```diff
@@ -127,12 +136,45 @@ internal set
         /// </remarks>
         public void SpinOnce()
         {
-            SpinOnce(DefaultSleep1Threshold);
+            SpinOnceCore(DefaultSleep1Threshold);
+        }
+
+        /// <summary>
+        /// Performs a single spin.
+        /// </summary>
+        /// <param name="sleep1Threshold">
+        /// A minimum spin count after which <code>Thread.Sleep(1)</code> may be used. Zero may be used to disable the use of
+        /// <code>Thread.Sleep(1)</code>.
```

Comment:
0 seems like a strange choice here.  If the argument is the minimum count above which Sleep(1) may be used, then using a smaller number increases the chance it'll be used, but then when we transition from 1 to 0, it makes it much less likely (i.e. impossible) it'll be used.  Would it be better to use int.MaxValue, since larger values make it less likely that Sleep(1) happens, and int.MaxValue is the largest possible value here?

Reply:
0 doesn't really make sense for this threshold because the intention is to just Sleep(1), and they may as well use Sleep(1) directly instead of using SpinWait, so I thought it could be reused. int.MaxValue is fine too, or similar to timeouts where -1 == no timeout. Maybe -1 would be good?

Reply:
> 0 doesn't really make sense for this threshold because the intention is to just Sleep(1), and they may as well use Sleep(1) directly instead of using SpinWait

Maybe, except that this argument is an argument to SpinOnce, not to the SpinWait ctor, which suggests that you might be using SpinWait with different values on each call (though I'm not sure in what situations you would do that), in which case if you wanted them all counted and wanted a Sleep(1), you could use 0.

I'd be fine with either int.MaxValue or -1.

Reply:
I'll go with -1 then. Do you think that SpinOnce(0) should Sleep(1)? I'm also adjusting the value to be at least YieldThreshold because otherwise NextSpinWillYield (which does not have knowledge about the threshold) may not be truthful.

Reply:
> Do you think that SpinOnce(0) should Sleep(1)?

I think it should be allowed to but it needn't be guaranteed to.  Basically, I would do whatever naturally falls out rather than special-casing it, if possible.  If that means that SpinOnce(0) will always Sleep(1), I think that makes sense and is fine.

Reply:
Ok, maybe I can just document that the threshold specified in this overload overrides what NextSpinWillYield says.

Reply:
or that it "may" override, that is NextSpinWillYield can't be trusted

Reply:
I feel it is better to leave it up to implementation. Currently the threshold for yielding can't be controlled, but suppose we were to allow that, it seems logical to say that that Sleep(1) threshold must be >= the yield threshold.

Reply:
In that case too, NextSpinWillYield would be meaningless

Reply:
Reverted that last commit. I think it makes sense to treat it as two separate phases, the spin-wait phase and yield/sleep phase. Since the yield threshold is not specified, it'll use the predefined value and the yield/sleep should only come after that. The main purpose of the parameter is to guarantee that sleep(1) is not used before reaching that count so that there is no unexpected long delay. Even after reaching the sleep(1) threshold the implementation may for instance choose to alternate sleep(1) with other types of yielding for better fairness.